src/table/memo.rs

```diff
@@ -22,6 +24,126 @@ pub(crate) struct MemoTable {
 pub trait Memo: Any + Send + Sync {
     /// Returns the `origin` of this memo
     fn origin(&self) -> &QueryOrigin;
+    fn clear_value(&mut self);
+}
+
+/// An untyped memo that can only be dropped.
+#[derive(Debug)]
+pub struct MemoDrop(NonNull<DummyMemo>, unsafe fn(NonNull<DummyMemo>));
```

Comment:
Three things:
1. I didn't know it was legal to write `unsafe fn(NonNull<DummyMemo>)`. I'm guessing that if you pass it a safe closure, this will fail to compile?
2. Can you make these named fields instead of tuple structs?
3. I *really* think this should be tested with Shuttle. if you're okay with me pushing to your branch, i can do that?

Reply:
> No, safe functions/closures can be coerced to unsafe ones just fine. After all calling a safe function in an unsafe context isn't problematic. Only the other direction won't work for obvious reasons.

The other two questions won't be relevant anymore